### PR TITLE
feat(activity): change to use conversations API

### PR DESCRIPTION
**Pull request by jonniespratley**
- Updated activities adapter to use `request` to fetch from conversations API.
- Updated response to handle parsing and formatting according to same API.

### CODE REVIEW

File: src/ActivitiesSDKAdapter.js [hunk 1]
```diff
@@ -30,11 +59,11 @@ import logger from './logger';
 function fromSDKActivity(sdkActivity) {
   return {
     ID: sdkActivity.id,
-    roomID: sdkActivity.roomId,
-    text: sdkActivity.text,
-    personID: sdkActivity.personId,
-    attachments: sdkActivity.attachments || [],
-    created: sdkActivity.created,
+    roomID: sdkActivity.target ? sdkActivity.target.id : sdkActivity.roomID,
```
Comment (jonniespratley) [inline src/ActivitiesSDKAdapter.js:62]: Here we need to convert to hydra id

File: src/ActivitiesSDKAdapter.js [hunk 2]
```diff
@@ -30,11 +59,11 @@ import logger from './logger';
...
+    text: sdkActivity.object ? sdkActivity.object.content : sdkActivity.text,
+    personID: sdkActivity.actor ? sdkActivity.actor.id : sdkActivity.personID,
```
Comment (jonniespratley) [inline src/ActivitiesSDKAdapter.js:64]: Need to convert `actor.id` to a `hydraID`

File: src/ActivitiesSDKAdapter.js (hunk 1 above)
Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js:62]: AKA `roomId` === `target.id`

File: src/ActivitiesSDKAdapter.js [hunk 3]
```diff
@@ -20,6 +20,35 @@ import logger from './logger';
...
+    }
+
+    // If object.cards array, then parse cards JSON string into array.
+    if (sdkActivity.object && sdkActivity.object.cards) {
+      const card = JSON.parse(sdkActivity.object.cards[0]);
+
+      attachments.push(card);
+    }
+
+    // If object.files, then merge files.items into attachments.
```
Comment (jonniespratley) [inline src/ActivitiesSDKAdapter.js:42]: Handle case where card and files.items exist.

File: src/ActivitiesSDKAdapter.js [hunk 4]
```diff
@@ -1,3 +1,5 @@
+import {deconstructHydraId} from '@webex/common';
+import { constructHydraId } from '@webex/common/dist/uuid-utils';
```
Comment (jonniespratley) [inline src/ActivitiesSDKAdapter.js:2]: Will fix

File: src/ActivitiesSDKAdapter.js [hunk 5]
```diff
@@ -29,12 +62,12 @@ import logger from './logger';
  */
 function fromSDKActivity(sdkActivity) {
   return {
-    ID: sdkActivity.id,
-    roomID: sdkActivity.roomId,
-    text: sdkActivity.text,
-    personID: sdkActivity.personId,
-    attachments: sdkActivity.attachments || [],
-    created: sdkActivity.created,
+    ID: constructHydraId('message', sdkActivity.id),
```
Comment (jonniespratley) [inline src/ActivitiesSDKAdapter.js:65]: One question, will it always be a non-hydra id? so we will always construct?

File: src/ActivitiesSDKAdapter.js [hunk 6]
```diff
@@ -20,6 +22,37 @@ import logger from './logger';
...
+      attachments.push(card);
+    }
+
+    // If object.files, then merge files.items into attachments.
+
+    return attachments;
+  } catch (err) {
+    logger.error('ACTIVITY', undefined, 'parseSDKAttachments()', `Unable parse attachments for activity with id "${sdkActivity.id}"`, err);
+
+    return [];
```
Comment (cipak) [inline src/ActivitiesSDKAdapter.js:52]: Hi, we have a single-return policy, maybe you could try-catch just around the json parse and return attachments at the end of the function.

Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js:54]: Single return policy. ok will do

File: src/ActivitiesSDKAdapter.js [hunk 7]
```diff
@@ -20,21 +21,54 @@ import logger from './logger';
...
+ *
+ * @private
+ * @param {object} sdkActivity  SDK activity object
+ * @returns {Array} Array of attachments.
+ */
+function parseSDKAttachments(sdkActivity) {
+  const {attachments = []} = sdkActivity;
+
+  if (sdkActivity.attachments) {
+    attachments.concat(sdkActivity.attachments);
```
Comment (cipak) [inline src/ActivitiesSDKAdapter.js:35]: won't this double the attachments extracted on line 32?

File: src/ActivitiesSDKAdapter.js [hunk 8]
```diff
@@ -20,21 +21,54 @@ import logger from './logger';
...
+  }
+
+  // If object.files, then merge files.items into attachments.
+  if (sdkActivity.card) {
+    attachments.push(sdkActivity.card);
+  }
+
+  try {
+    if (sdkActivity.object && sdkActivity.object.cards) {
+      const card = JSON.parse(sdkActivity.object.cards[0]);
```
Comment (cipak) [inline src/ActivitiesSDKAdapter.js:45]: from our tests, the cards are encrypted in the sdk activity object. They need to be decrypted.

Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js:45]: Ok, how would i go about doing this?

File: src/ActivitiesSDKAdapter.js (hunk 7 above)
Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js:35]: Nice one, will fix.

File: src/ActivitiesSDKAdapter.js (hunk 8 above)
Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js:40]: After investigation, I did not see cards or attachments being returned encrypted from the API.

**Review by taymoork2 — APPROVED** [review]
LGTM

File: src/ActivitiesSDKAdapter.js [hunk 9]
```diff
@@ -20,21 +21,51 @@ import logger from './logger';
...
+
 /**
  * Maps SDK activity to adapter activity
  *
  * @private
  * @param {object} sdkActivity  SDK activity object
  * @returns {Activity} Adapter activity object
  */
-function fromSDKActivity(sdkActivity) {
+export function fromSDKActivity(sdkActivity) {
```
Comment (lalli-flores) [inline src/ActivitiesSDKAdapter.js:56]: No need to export private methods
```suggestion
function fromSDKActivity(sdkActivity) {
```

File: src/ActivitiesSDKAdapter.js [hunk 10]
```diff
@@ -20,21 +21,51 @@ import logger from './logger';
...
+ * Extracts JSON card and files from server activity.
+ *
+ * @private
+ * @param {object} sdkActivity  SDK activity object
+ * @returns {Array} Array of attachments.
+ */
+function parseSDKAttachments(sdkActivity) {
+  const {attachments = []} = sdkActivity;
+
+  if (sdkActivity.card) {
```
Comment (lalli-flores) [inline src/ActivitiesSDKAdapter.js]: Not sure understand this condition here

File: src/ActivitiesSDKAdapter.js [hunk 11]
```diff
@@ -20,21 +21,51 @@ import logger from './logger';
...
+  try {
+    if (sdkActivity.object && sdkActivity.object.cards) {
+      sdkActivity.object.cards.forEach((c) => {
+        const card = JSON.parse(c);
+
+        attachments.push(card);
+      });
+    }
+  } catch (err) {
+    logger.error('ACTIVITY', undefined, 'parseSDKAttachments()', `Unable parse attachments for activity with id "${sdkActivity.id}"`, err);
```
Comment (lalli-flores) [inline src/ActivitiesSDKAdapter.js:43]: If you're handling the error, no need to report it as an error unless you're throwing it. I would make this a warning and specify in the log the error is being handlded

Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js]: Done

File: src/ActivitiesSDKAdapter.js (hunk 10 above)
Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js]: Removed

File: src/ActivitiesSDKAdapter.js (hunk 9 above)
Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.js]: removed

File: src/ActivitiesSDKAdapter.test.js [hunk 12]
```diff
@@ -96,60 +96,29 @@ describe('Activities SDK Adapter', () => {
...
+          },
+          published: created,
+        }),
+      );
+
       activitiesSDKAdapter.postActivity(activityData).pipe(last()).subscribe((activity) => {
         expect(activity).toMatchObject({
-          ID: 'activityID',
-          roomID: 'roomID',
+          ID,
```
[the remark this answers is not in the document]
Comment (jonniespratley) [reply src/ActivitiesSDKAdapter.test.js:119]: I thought the transform handles making a server activity. `104` to correct structure.

**Review by lalli-flores — APPROVED** [review]

**Pull request merged**